rfcs/20190116-embedding-partitioned-variable.md

```diff
@@ -0,0 +1,498 @@
+              splits=None):
+    """Specification returned from a partitioner.
+
+    Args:
+      partitions: List of integers. Number of partitions for each dimension.
+      splits: List of lists of ints, indicating how each dimension can be splitted,
+        allowing uneven sharding. Either partitions or splits should be given but
+        not both.
+    """
+    return super(PartitionSpec, cls).__new__(
+        cls,                                                                                    
+        partitions=partitions,                                                                                                                                                  
+        splits=splits)
+```
+
+This `PartitionSpec` can be used in all Distribution Strategies.
+
+
+#### Explore Partitioning in CollectiveAllReduceStrategy
+
+
+On multiple hosts with `CollectiveAllReduceStrstegy` we can experiment with sharding a large embedding across different workers. This may require an efficient all-to-all personalized communication and a network with smaller start-up time.
+
+
+#### Explore Hybrid Strategy
+
+
+We can experiment with a hybrid strategy that mirrors some variables and shards other variables. All the ops can still be mirrored on each replica. We can have a strategy like:
+
+```python
+class HybridStrategy(DistributionStrategy):
+  def __init__(strategy1, strategy2, strategy_chooser):
+    """A hybrid distribution strategy.
+
+    Args:
+      strategy1: the main default strategy object.
+      strategy2: another strategy object.
+      strategy_chooser: a callable that takes variable name, variable shape and
+        variable dtype and returns which strategy to apply for this variable.
+    """
+    pass
+```
+
+
+# Alternatives Considered
+
+
+### Embedding in Mirrored Architecture: Sharding or Mirroring?
+
+
+In the `MirroredStrategy` or `CollectiveAllReduceStrategy`, there are several ways of handling embeddings. However due to the fact that `dynamic_partition` and `dynamic_stitch` required by embedding lookups on sharded embeddings only have CPU kernels, we can not shard embeddings across GPUs.
+
+We'll discuss how embeddings will be handled in single host and multi-host cases.
+
+
+##### Single host
+
+
+On a single host, embeddings can be placed on host memory or mirrored on devices.
+
+When an embedding variable is small, it may benefit from being mirrored on devices. Lookup can be performed local to each device. Updates can be casted to dense tensors and use the existing allreduce primitive to exchange gradients. This can still be faster than running allgather on their corresponding sparse updates.
+
+When an embedding variable is larger, we'll need an efficient allgather primitive to exchange updates between devices. Alternatively, we can place it on host memory at the cost of transferring embeddings from host to devices in the forward pass and gathering updates from devices to host in the backward pass.
```

Comment:
Have we considered for embedding partition using GPU memory rather than host memory? This may bring  more complexity, but may also bring more performance gains since host2device traffic can be reduced. 
Also, it looks that many embedding-dominating models are not complex so that GPU device memory may be a potential choice for holding the embedding partition.
Actually, inside Alibaba, we are working on prototyping a new strategy called as PEARLStrategy(Partitioned Embedding And RepLicated variables ) to support GPU-based embedding partitioning.

Reply:
I've just updated these paragraphs. Sharding on GPUs might be faster but more difficult to implement. We'll implement mirroring embeddings and allgather first.

I am not sure how faster it is to shard across GPUs than to put it on host memory. You've probably done some experiment : ). If it worths the complexity, we'll definitely implement it.